doc/wapiti.1
.TH wapiti 1

.SH NAME
wapiti

.SH SYNOPSIS
.B wapiti
.RB mode\ [options]\ [input]\ [output]

.SH DESCRIPTION

.SS Overview
Wapiti is a program for training and using discriminative sequence labelling models with various algorithms using an elastic penalty.
It currently implements maximum entropy models, maximum entropy Markov models (MEMM) and linear-chain conditional random field (CRF) models.
.P
Depending on the mode argument, it can train a model, label new data, or dump a model in readable form.
.P
The mode argument can be "train", "label", "dump", or "update". (Only a prefix long enough to differentiate them is really needed.)

.SS Options
.TP
.B \-h | \-\-help
Output a short help message.
.TP
.B \-\-version
Output version and revision numbers.

.SS Train mode
.TP
.B \-\-me
Activate the pure maxent mode, see below for more details.
.TP
.B \-T | \-\-type <string>
Select the type of model to train, "maxent", "memm", or "crf". Use "list" to get a list of supported models types. The default is to train CRF models.
.TP
.B \-a | \-\-algo <string>
Select the algorithm used to train the model. Use "list" for a list of available algorithms. The first algorithm in this list is used as default.
.TP
.B \-p | \-\-pattern <file>
Specify the file containing the patterns for extracting features. The format of the pattern file is detailed below.
.TP
.B \-m | \-\-model <file>
Specify a model file to load and to train again. This allows you either to continue an interrupted training or to use an old model as a starting point for training with new data. Beware that no new labels can be inserted in the model. As the training parameters are not saved in the model file, you have to specify them again. Specify new training parameters if, for example, you want to continue training with a different algorithm or penalty.
.TP
.B \-d | \-\-devel <file>
Specify the data file to load as a development set. At the end of each iteration the error rate is computed on this dataset and displayed in the progress line. If enabled, this value is used to check convergence and stop training. If no development file is specified, the training set is used instead, but beware that it is bad practice to use the training set to choose the stopping criterion.
.TP
.B \-\-rstate <file>
Restore an optimizer state from the given file and restart optimization from this point. This option is only available for the L-BFGS and R-PROP algorithms, but the saved state is compatible between MEMM and CRF models. This allows more information about the optimal point found while training an MEMM to be kept, to bootstrap a CRF model or to restart an optimization with adjusted parameters.
.TP
.B \-\-sstate <file>
Save the full optimizer state at the end of optimization so it can be restored later with \-\-rstate.
.TP
.B \-c | \-\-compact
Enable model compaction at the end of the training. This will remove all inactive observations from the model, leading to a much smaller model when an l1-penalty is used. See the note below for more details.
.TP
.B \-t | \-\-nthread <integer>
Set the number of threads to use. This can drastically improve performance---depending on the algorithm. The best value is the number of cores on your CPU. The default is 1.
.TP
.B \-j | \-\-jobsize <integer>
Set the size of the job a thread will get each time it has nothing more to do. This is the number of sequences to proceed. The default is 64. Increasing it will reduce communication overhead but can lead to a bad balancing between threads. Reducing it increases the communication overhead but can balance work better between threads in case the dataset is small.
.B \-s | \-\-sparse
Enable the computation of the forward/backward in sparse mode.
.TP
.B \-i | \-\-maxiter <integer>
Defines the maximum number of iterations done by the training algorithm. A value of 0 means training will continue until another stopping criterion is reached. The default is 0.
.TP
.B \-1 | \-\-rho1 <float>
Defines the L1-component of the elastic-net penalty. Increasing this value leads to smaller models and can improve training time but will probably lead to reduced performances. Setting this value to 0 results in a classical L2-penalty only. If the algorithm can optimize the L1-penalty, the default value is 0.5, otherwise the default is 0.0.
.TP
.B \-2 | \-\-rho2 <float>
Specifies the L2-component of the elastic-net penalty. Setting this value to 0 leads to simple L1 regularization. While allowed, this is discouraged as it can lead to numerical instability. The default value is 0.00001.
.TP
.B \-o | \-\-objwin <integer>
Set the window size for the objective value stopping criterion. See below for more details. Default value is 5.
.TP
.B \-w | \-\-stopwin <integer>
Set the window size for the development stopping criterion. See below for more details. Default value is 5.
.TP
.B \-e | \-\-stopeps <float>
Set the size of the interval for stopping criterion, see below for more details. Default value is 0.02%.
.TP
.B \-\-clip
Enables gradient clipping for the L-BFGS. This will set to 0 the gradient component whose corresponding feature values are 0, preventing the trainer from moving the feature away from 0. This is useful if you have a sparse model and want to refine it with an l2-only regularization without losing the sparsity.
.TP
.B \-\-histsz <integer>
Specifies the size of the history to keep in L-BFGS to approximate the inverse of the diagonal of the Hessian. Increasing this value leads to better approximation, and therefore generally fewer iterations but increased memory usage. The default is 5.
.TP
.B \-\-maxls <integer>
Set the maximum number of linesearch steps in L-BFGS to perform before giving up.
.TP
.B \-\-eta0 <float>
Set the learning rate for the SGD trainer.
.TP
.B \-\-alpha <float>
Set the alpha value of the exponential decay in the SGD trainer.
.TP
.B \-\-kappa <float>
Set the kappa parameter for the BCD trainer. Default is 1.5. Increasing this value makes the algorithm more stable but also slower. Try to increase it if you have numerical instability.
.TP
.B \-\-stpmin <float>
.B \-\-stpmax <float>
Set the minimum/maximum step size allowed for the RPROP trainer. Defaults are 1.0e-8 and 50.0. The defaults seem to get numerical stability with double computations.
.TP
.B \-\-stpinc <float>
.B \-\-stpdec <float>
Set the increment/decrement factor used to update the steps in the RPROP trainer. Default values are 1.2 and 0.5. Increment must be greater than 1.0 and decrement must be between 0 and 1.0.
.TP
.B \-\-cutoff
Select the alternate projection scheme for RPROP with l1-regularization. This can lead to a better model depending on your task.

.SS Label mode
.TP
.B \-\-me
Activate the pure maxent mode, see below for more details.
.TP
.B \-m | \-\-model <file>
Specifies a model file to load and to use for labeling. This switch is mandatory.
.TP
.B \-l | \-\-label
With this switch, Wapiti will only output the predicted labels. Without, it will output the full data with an additional column containing the predicted labels.
.TP
.B \-c | \-\-check
Assume the data to be labeled are already labeled so during the labeling process we can check our own result, displaying the error rates. This doesn't affect the labeling process---output data will remain exactly the same. However, progress will be more verbose and informative: at the end of the process, for each labels, the precision, recall, and f-measure will be displayed. If you ask for N-best output, statistics are computed only on the best sequence.
.TP
.B \-s | \-\-score
Output a line with score before the data. The line start with a '#' symbol followed by the output number in the n-best list and the score of the sequence of labels. Also output a score for each label of the sequence. Beware that if you use viterbi labelling, this is a raw score, not really meaningful. It is not normalized so it cannot be interpreted as a probability. To get normalized scores, you must use posterior decoding.
.TP
.B \-p | \-\-post
Use posterior decoding instead of the standard Viterbi decoding. This generally produces better results, at the cost of a slower decoding. This also allows users to output normalized scores for sequences and labels.
.TP
.B \-n | \-\-nbest <int>
Output the N-best sequences of labels instead of just the best one. The N sequences of labels are generated  in the output file in decreasing order of their score (best hypothesis first).
.TP
.B \-\-force
Enable forced decoding for labeling sequences that are already partially labeled. See below for details.

.SS Dump mode
.TP
.B \-p | \-\-prec <int>
Set the floating point precision of weights values.
.TP
.B \-\-all
Force dumping of all features---even the zero ones.

.SS Update mode
.TP
.B \-m | \-\-model <file>
Specifies the model file to load and to update with the correction from the input file.
.TP
.B \-c | \-\-compact
Force removal of blocks of zero features before saving the updated model file.

.SH USAGE
Wapiti can work in different modes. The mode determines the options that are available (see above) and what the model expects in the input and output files. In train mode, Wapiti expects a training dataset as input and outputs the trained model. In label mode, it expects data to label as input and will output the same data, augmented with the labels computed by the model. Finally, in dump mode, it expects a model as input and outputs it in a readable form.
.P
In train mode, Wapiti will load an existing model if one is given, will read the train dataset as well as an optional development set, and will estimate the model. Progress information is output during all these steps. Training stops either when the model is fully optimized or when one of the stopping criteria is reached or when the user sends a TERM signal. (see below)
.P
In label mode, progress is not very informative except when the user supplies data with ground truth labels. In this case, error rates will be computed and reported.

.SH STOPPING CRITERIA
.P
There are various ways to stop training, depending on the command line switch provided.
.P
The simplest criterion is the iteration count. By default, algorithms will iterate forever, but you can specify a maximum number of iterations with \-\-maxiter.

Finding the exact optimum is generally not needed to get the best model. There is an infinity number of points around the optimum that lead to a model as good as the best one. The error window criterion checks for this by looking at the error rate of the model over the development set and stopping training when it is stable enough. To do this, the error rate of the last few iterations is kept. When the difference between extreme values falls bellow a given value, training is stopped. (If no development set is given, the error rates are computed over the training data, but this is bad practice.)

For algorithms that provide the objective function value at each iteration, we also stop them when this value has not changed significantly over the past few iterations. This window size is controlled by the objwin parameter.

Each algorithm can also provide its own stopping system. For example, l-bfgs stops when numerical precision prevents further progress.

The last criterion is the user. By sending a TERM signal to Wapiti you instruct it to stop training as soon as possible, discarding the last computation, in order to finish training and save the model. If you don't care about the model, sending a second TERM signal will make the program violently exit without saving anything. (On most systems, a TERM signal can be send with CTRL-C.)

.SH REGULARIZATION
.P
Wapiti uses the elasitc-net penalty of the form
.TP
rho_1 * |theta|_1 + rho_2 / 2.0 * ||theta||_2^2
.P
This means that you can choose to use the full elastic-net or more classical L1 or L2 penalty. To fall back to one of these, you just have to set, respectively, rho1 or rho2 to 0.0.

Some algorithms work only with one or the other component. In this case, the value of the other is simply ignored. See the documentation pertaining to each specific algorithm for more details.

.SH ALGORITHMS
.B l-bfgs
This is the classical quasi-Newton optimization algorithm with limited memory. It works by approximating the inverse of the diagonal Hessian using a history of the previous values of the feature weights and of the gradient.

This algorithm requires the gradient to be fully computable at any point, so it cannot do L1 regularization. In this case, the OWL-QN variant is used instead, enabling the use of the full elastic-net penalty.

It requires 5 + M * 2 vectors to be kept, the sizes of which are the number of features. Each component of these vectors is a double precision floating point value. So, for training a model with F features, you need 8 * F * (5 + M * 2) bytes of memory. If the OWL-QN variant is used, one additional vector is needed to keep the pseudo-gradient.

.B sgd-l1
This is the stochastic gradient descent algorithm for an L1-regularized model. It works by computing the gradient on a single sequence at a time and making a small step in this direction.

The SGD algorithm will very quickly find an acceptable solution for the model, but will take a longer time to find the optimal one, and there is no guarantee it will ever find it.

The memory requirements are lighter than for quasi-Newton methods as it requires only 3 vectors, the sizes of which are the number of features.

.B bcd
This is the blockwise coordinate descent algorithm with elastic-net penalty. This algorithm is best suited for very large label sets and sparse feature sets. It optimizes the model one observation at a time, going through all observations at each iteration. It usually converges in only a few dozen iterations (rarely more than 30).

This is the more memory-economical algorithm as it only requires the feature weight vector to be kept in memory. In this algorithm, using complex bigram features comes almost for free.

This flexibility has a price: don't use it if your features are not sparse, as it will be very slow in this case.

NOTE: This algorithm is available only for training CRF models.

.B rprop (rprop+ / rprop-)
This algorithm uses the gradient only to find a good search direction, not for choosing the step to make in that direction. It can be very effective on some datasets.

Compared to quasi-newton methods, rprop reaches the neighboorhood of the optimum much more quickly, but the lack of second order information and the restricted use of the first order information makes the fine tuning slower.

Memory requirements are quite light as this algorithm only requires 4 vectors of the size of the feature set.

The rprop- is a variant of rprop+ without backtracking. Its performance compared to rprop+ is task dependent and it requires one less vector; so for very large models it can be better to use this option than the standard approach.

.SH MULTI-THREADING
Wapiti can efficiently use multiple threads to accelerate the gradient computation for l-bfgs and rprop algorithms. Using the --nthread parameter, you can specify the number of threads to use.

Beware that if atomic updates were disabled at compilation time, each thread after the first will cost you an extra vector of the size of the feature set. This implies that for large models, multiple threads can cost you a lot of memory. Atomic updates are supported at least with GCC and CLang compilers. It may also work if your compiler supports the same intrinsic atomic operations, or if you reimplement the atm_inc function in gradient.c for it.

The multi-threading code can be disabled at compilation time if your platform does not support it. See wapiti.h for more details.

.SH DATAFILES
Data files are plain text files containing sequences separated by empty lines. Each sequence is a set of non-empty lines. Each line represents one position in the sequence.

Each line is made of tokens separated either by spaces or by tabs. All tokens are observations available for training or labeling, except for the last one: in training mode, the last token is assumed to be the label to predict.

If no pattern is specified, each token is interpreted directly as an observation and is combined with the label in order to generate features. If patterns are specified, they are used in combination with the tokens to generate the features. Each observation must be prefixed by 'u', 'b' or '*' in order to specify whether it is unigram, bigram or both.

.SH PATTERNS
Pattern files are almost compatible with CRF++ templates. Empty lines as well as all characters appearing after a '#' are discarded. The remaining lines are interpreted as patterns.

The first char of a pattern must be 'u', 'b' or '*' (in upper or lower case). This indicates the type of features that will be generated from this pattern: respectively unigram, bigrams or both.

The remaining part of the pattern is used to build an observation string. Each marker of the kind "%x[off,col]" is replaced by the token in the column "col" from the data file at current position plus the offset "off".
The "off" value can be prefixed with an "@" to make it an absolute position from the start of the sequence (if positive) or from the end (if negative). An offset of "@1" will thus refer to the first symbol of the current sequence and "@-1" to the last one.

For example, if your data is:
    a1    b1    c1
    a2    b2    c2
    a3    b3    c3
.br
The pattern "u:%x[-1,0]/%x[+1,2]" applied at position 2 in the sequence will produce the observation "u:a1/c3".

Note that sequences are implicitely padded with special tokens such as "_X-1" or "_X+2" in order to apply markers with arbitrary offset at any position in the sequence. This means, for instance, that "_X-1" denotes the left context of the first token in a sequence.

Wapiti also supports a simple kind of matching that can be useful, for example, in natural language processing applications. This is done using two other commands of the form %m[off,col,"regexp"] and %t[off,col,"regexp"]. Both commands will get data the same way as the %x command using the "col" and "off" values but will apply a regular expression to the data before substituting it. The %t will replace the data by "true" or "false" depending on whether the expression matches on the data or not. The %m command replaces the data by the substring matched by the expression.

The regular expression matcher is a subset of those found in classical unix system but is sufficient for most tasks. The recognized subset is quite simple. First for matching characters:
     .  -> match any character
     \\x -> match a character class (in uppercase, match the complement)
             \\d : digit       \\a : alpha      \\w : alpha + digit
             \\l : lowercase   \\u : uppercase  \\p : punctuation
             \\s : space
           or escape a character
     x  -> any other character matches itself
.br
And the constructs :
     ^  -> at the beginning of the regexp, anchor it at start of string
     $  -> at the end of the regexp, anchor it at end of string
     *  -> match any number of repetitions of the previous character
     ?  -> optionally match the previous character
So, for example, the regexp "^.?.?.?.?" will match a prefix of at most four characters and "^\u\u*$" will match only on data composed solely of uppercase characters.

For the commands %x, %t, and %m, if the command name is given in uppercase, the case is removed from the string before being added to the observation.

.SH FORCED DECODING
The forced decoding switch enables decoding partially-labelled data. If some labels are already known and only the unknown ones must be predicted, instead of doing a full prediction and correcting the Wapiti output as a post-processing step, it is possible to enforce decoding. This allows you to specify the already known labels and let Wapiti use this information to improve the decoding.


HISTORY
18/12/2013
  Release v1.5.0: Update mode and bug fixes

    Add precision specifier for model dumping
    Add update mode to modify a model
    Lots of english corrections in the manual
    Fix bug in model format compatibility
    Fix memory allocation with large models
    Fix small memory bug in quark database
    Fix bug with bigram features in raw mode

23/04/2012
  Release v1.4.0: Forced decoding, optimizer state, and bug fixes

    Add forced decoding to partialy decode sequences
    Add optimizer state saving for L-BFGS and R-PROP
    Switched to elapsed time instead of wall time in progress
    Fix bug in Makefile (thanks to Lars Buitinck)
    Fix local normalization decoding in MEMM (thanks to Anoop Deoras)
    Fix bad handling of objwin option
    Fix bug in reader for single letter obs and lbl

03/11/2011
  Release v1.3.0: MEMM, faster gradient and bug fixes

    Added support for Maximum Entropy Markov Models.
    Added use of atomic operation in gradient computation
    Improved RProp numerical stability
    Fix bug with unseen features in raw mode (thanks to Jurgen Van Gael)
    Fix bug discarding some features in maxent mode (thanks to George Foster)
    Switched code to stdint, should resolve some issue with size_t
      on exotic systems.

07/07/2011
  Release v1.2.0: RProp improvements and bug fix

    Switch from splay-trees to critbit-tries
    Add RProp+ and RProp- variants of the RProp algorithm
    Add a new projection scheme for RProp with l1
    Make maxent work with sequences
    Add space matching in regexp
    Fix a few small bugs

19/03/2011
  Release v1.1.3: Some small improvements

    New option --jobsize for fine grained multi threading
    Improved SGD index construction : a lot faster
    Fix a small bug in sparse multi-threaded gradient

12/11/2010
  Release v1.1.2: Bug fix release

    Fix a small bug in L-BFGS/OWL-QN, should improve a bit
      convergence speed in some case.
    Fix a bug in multi-thread job system thanks to Alexander Fraser,
      should fix error rates and training speed on large dataset.
    Fix two small memory leaks.
    Some improvment in quark database handling.

24/09/2010
  Release v1.1.1: Mainly multi-threading improvements

    RPROP algorithm is now fully multi-threaded.
    Error rate estimation during training is now multi-threaded.
    Better jobs scheduling in multi-threaded gradient.
    Multi-threading code can be disabled (compilation on Windows should
      be simpler).
    Fixed bug in L1 optimization with RPROP (should improve stability).

08/09/2010
  Release v1.1.0: A few new features

    Added maxent mode.
    Added decoding through posteriors, this should improve accuracy
      at the price of computational time.
    Added the RPROP optimization algorithm.
    Added absolute indexing in patterns.
    Changed the scored output format as the posterior decoding
      provide normalized score at each position. The output is now
      compatible with CRF++.
      Some code cleanup.

29/07/2010
  Release v1.0.2: Mainly a bug fix version.

    Fixed some memory leaks, thanks to David Keeler
    Fixed argument processing to be more user friendly
    Fixed small bug in model compaction
    Added reading of raw files
    Spell corrections in man page

18/06/2010
  Release v1.0.0: Initial public version.



src/wapiti.c
/*
 *      Wapiti - A linear-chain CRF tool
 *
 * Copyright (c) 2009-2013  CNRS
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
#include <ctype.h>
#include <inttypes.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "decoder.h"
#include "model.h"
#include "options.h"
#include "progress.h"
#include "quark.h"
#include "reader.h"
#include "sequence.h"
#include "tools.h"
#include "trainers.h"
#include "wapiti.h"

/*******************************************************************************
 * Training
 ******************************************************************************/
static const char *typ_lst[] = {
	"maxent",
	"memm",
	"crf"
};
static const uint32_t typ_cnt = sizeof(typ_lst) / sizeof(typ_lst[0]);

static const struct {
	char *name;
	void (* train)(mdl_t *mdl);
} trn_lst[] = {
	{"l-bfgs", trn_lbfgs},
	{"sgd-l1", trn_sgdl1},
	{"bcd",    trn_bcd  },
	{"rprop",  trn_rprop},
	{"rprop+", trn_rprop},
	{"rprop-", trn_rprop},
};
static const uint32_t trn_cnt = sizeof(trn_lst) / sizeof(trn_lst[0]);

static void dotrain(mdl_t *mdl) {
	// Check if the user requested the type or trainer list. If this is not
	// the case, search them in the lists.
	if (!strcmp(mdl->opt->type, "list")) {
		info("Available types of models:\n");
		for (uint32_t i = 0; i < typ_cnt; i++)
			info("\t%s\n", typ_lst[i]);
		exit(EXIT_SUCCESS);
	}
	if (!strcmp(mdl->opt->algo, "list")) {
		info("Available training algorithms:\n");
		for (uint32_t i = 0; i < trn_cnt; i++)
			info("\t%s\n", trn_lst[i].name);
		exit(EXIT_SUCCESS);
	}
	uint32_t typ, trn;
	for (typ = 0; typ < typ_cnt; typ++)
		if (!strcmp(mdl->opt->type, typ_lst[typ]))
			break;
	if (typ == typ_cnt)
		fatal("unknown model type '%s'", mdl->opt->type);
	mdl->type = typ;
	for (trn = 0; trn < trn_cnt; trn++)
		if (!strcmp(mdl->opt->algo, trn_lst[trn].name))
			break;
	if (trn == trn_cnt)
		fatal("unknown algorithm '%s'", mdl->opt->algo);
	// Load a previous model to train again if specified by the user.
	if (mdl->opt->model != NULL) {
		info("* Load previous model\n");
		FILE *file = fopen(mdl->opt->model, "r");
		if (file == NULL)
			pfatal("cannot open input model file");
		mdl_load(mdl, file);
	}
	// Load the pattern file. This will unlock the database if previously
	// locked by loading a model.
	if (mdl->opt->pattern != NULL) {
		info("* Load patterns\n");
		FILE *file = fopen(mdl->opt->pattern, "r");
		if (file == NULL)
			pfatal("cannot open pattern file");
		rdr_loadpat(mdl->reader, file);
		fclose(file);
		qrk_lock(mdl->reader->obs, false);
	}
	// Load the training data. When this is done we lock the quarks as we
	// don't want to put in the model, informations present only in the
	// devlopment set.
	info("* Load training data\n");
	FILE *file = stdin;
	if (mdl->opt->input != NULL) {
		file = fopen(mdl->opt->input, "r");
		if (file == NULL)
			pfatal("cannot open input data file");
	}
	mdl->train = rdr_readdat(mdl->reader, file, true);
	if (mdl->opt->input != NULL)
		fclose(file);
	qrk_lock(mdl->reader->lbl, true);
	qrk_lock(mdl->reader->obs, true);
	if (mdl->train == NULL || mdl->train->nseq == 0)
		fatal("no train data loaded");
	// If present, load the development set in the model. If not specified,
	// the training dataset will be used instead.
	if (mdl->opt->devel != NULL) {
		info("* Load development data\n");
		FILE *file = fopen(mdl->opt->devel, "r");
		if (file == NULL)
			pfatal("cannot open development file");
		mdl->devel = rdr_readdat(mdl->reader, file, true);
		fclose(file);
	}
	// Initialize the model. If a previous model was loaded, this will be
	// just a resync, else the model structure will be created.
	if (mdl->theta == NULL)
		info("* Initialize the model\n");
	else
		info("* Resync the model\n");
	mdl_sync(mdl);
	// Display some statistics as we all love this.
	info("* Summary\n");
	info("    nb train:    %"PRIu32"\n", mdl->train->nseq);
	if (mdl->devel != NULL)
		info("    nb devel:    %"PRIu32"\n", mdl->devel->nseq);
	info("    nb labels:   %"PRIu32"\n", mdl->nlbl);
	info("    nb blocks:   %"PRIu64"\n", mdl->nobs);
	info("    nb features: %"PRIu64"\n", mdl->nftr);
	// And train the model...
	info("* Train the model with %s\n", mdl->opt->algo);
	uit_setup(mdl);
	trn_lst[trn].train(mdl);
	uit_cleanup(mdl);
	// If requested compact the model.
	if (mdl->opt->compact) {
		const uint64_t O = mdl->nobs;
		const uint64_t F = mdl->nftr;
		info("* Compacting the model\n");
		mdl_compact(mdl);
		info("    %8"PRIu64" observations removed\n", O - mdl->nobs);
		info("    %8"PRIu64" features removed\n", F - mdl->nftr);
	}
	// And save the trained model
	info("* Save the model\n");
	file = stdout;
	if (mdl->opt->output != NULL) {
		file = fopen(mdl->opt->output, "w");
		if (file == NULL)
			pfatal("cannot open output model");
	}
	mdl_save(mdl, file);
	if (mdl->opt->output != NULL)
		fclose(file);
	info("* Done\n");
}

/*******************************************************************************
 * Labeling
 ******************************************************************************/
static void dolabel(mdl_t *mdl) {
	// First, load the model provided by the user. This is mandatory to
	// label new datas ;-)
	if (mdl->opt->model == NULL)
		fatal("you must specify a model");
	info("* Load model\n");
	FILE *file = fopen(mdl->opt->model, "r");
	if (file == NULL)
		pfatal("cannot open input model file");
	mdl_load(mdl, file);
	// Open input and output files
	FILE *fin = stdin, *fout = stdout;
	if (mdl->opt->input != NULL) {
		fin = fopen(mdl->opt->input, "r");
		if (fin == NULL)
			pfatal("cannot open input data file");
	}
	if (mdl->opt->output != NULL) {
		fout = fopen(mdl->opt->output, "w");
		if (fout == NULL)
			pfatal("cannot open output data file");
	}
	// Do the labelling
	info("* Label sequences\n");
	tag_label(mdl, fin, fout);
	info("* Done\n");
	// And close files
	if (mdl->opt->input != NULL)
		fclose(fin);
	if (mdl->opt->output != NULL)
		fclose(fout);
}

/*******************************************************************************
 * Dumping
 ******************************************************************************/
static void dodump(mdl_t *mdl) {
	// Load input model file
	info("* Load model\n");
	FILE *fin = stdin;
	if (mdl->opt->input != NULL) {
		fin = fopen(mdl->opt->input, "r");
		if (fin == NULL)
			pfatal("cannot open input data file");
	}
	mdl_load(mdl, fin);
	if (mdl->opt->input != NULL)
		fclose(fin);
	// Open output file
	FILE *fout = stdout;
	if (mdl->opt->output != NULL) {
		fout = fopen(mdl->opt->output, "w");
		if (fout == NULL)
			pfatal("cannot open output data file");
	}
	// Dump model
	info("* Dump model\n");
	const uint32_t Y = mdl->nlbl;
	const uint64_t O = mdl->nobs;
	const qrk_t *Qlbl = mdl->reader->lbl;
	const qrk_t *Qobs = mdl->reader->obs;
	char fmt[16];
	sprintf(fmt, "%%.%df\n", mdl->opt->prec);
	for (uint64_t o = 0; o < O; o++) {
		const char *obs = qrk_id2str(Qobs, o);
		bool empty = true;
		if (mdl->kind[o] & 1) {
			const double *w = mdl->theta + mdl->uoff[o];
			for (uint32_t y = 0; y < Y; y++) {
				if (!mdl->opt->all && w[y] == 0.0)
					continue;
				const char *ly = qrk_id2str(Qlbl, y);
				fprintf(fout, "%s\t#\t%s\t", obs, ly);
				fprintf(fout, fmt, w[y]);
				empty = false;
			}
		}
		if (mdl->kind[o] & 2) {
			const double *w = mdl->theta + mdl->boff[o];
			for (uint32_t d = 0; d < Y * Y; d++) {
				if (!mdl->opt->all && w[d] == 0.0)
					continue;
				const char *ly  = qrk_id2str(Qlbl, d % Y);
				const char *lyp = qrk_id2str(Qlbl, d / Y);
				fprintf(fout, "%s\t%s\t%s\t", obs, lyp, ly);
				fprintf(fout, fmt, w[d]);
				empty = false;
			}
		}
		if (!empty)
			fprintf(fout, "\n");
	}
	if (mdl->opt->output != NULL)
		fclose(fout);
}


/*******************************************************************************
 * Updating
 ******************************************************************************/
static void doupdt(mdl_t *mdl) {
	// Load input model file
	info("* Load model\n");
	if (mdl->opt->model == NULL)
		fatal("no model file provided");
	FILE *Min = fopen(mdl->opt->model, "r");
	if (Min == NULL)
		pfatal("cannot open model file %s", mdl->opt->model);
	mdl_load(mdl, Min);
	fclose(Min);
	// Open patch file
	info("* Update model\n");
	FILE *fin = stdin;
	if (mdl->opt->input != NULL) {
		fin = fopen(mdl->opt->input, "r");
		if (fin == NULL)
			pfatal("cannot open update file");
	}
	int nline = 0;
	while (!feof(fin)) {
		char *raw = rdr_readline(fin);
		if (raw == NULL)
			break;
		char *line = raw;
		nline++;
		// First we split the line in space separated tokens. We expect
		// four of them and skip empty lines.
		char *toks[4];
		int ntoks = 0;
		while (ntoks < 4) {
			while (isspace(*line))
				line++;
			if (*line == '\0')
				break;
			toks[ntoks++] = line;
			while (*line != '\0' && !isspace(*line))
				line++;
			if (*line == '\0')
				break;
			*line++ = '\0';
		}
		if (ntoks == 0) {
			free(raw);
			continue;
		} else if (ntoks != 4) {
			fatal("invalid line at %d", nline);
		}
		// Parse the tokens, the first three should be string maping to
		// observations and labels and the last should be the weight.
		uint64_t obs = none, yp = none, y = none;
		obs = qrk_str2id(mdl->reader->obs, toks[0]);
		if (obs == none)
			fatal("bad on observation on line %d", nline);
		if (strcmp(toks[1], "#")) {
			yp = qrk_str2id(mdl->reader->lbl, toks[1]);
			if (yp == none)
				fatal("bad label <%s> line %d", toks[1], nline);
		}
		y = qrk_str2id(mdl->reader->lbl, toks[2]);
		if (y == none)
			fatal("bad label <%s> line %d", toks[2], nline);
		double wgh = 0.0;
		if (sscanf(toks[3], "%lf", &wgh) != 1)
			fatal("bad weight on line %d", nline);

		const uint32_t Y = mdl->nlbl;
		if (yp == none) {
			double *w = mdl->theta + mdl->uoff[obs];
			w[y] = wgh;
		} else {
			double *w = mdl->theta + mdl->boff[obs];
			w[yp * Y + y] = wgh;
		}
		free(raw);
	}
	if (mdl->opt->input != NULL)
		fclose(fin);
	// If requested compact the model.
	if (mdl->opt->compact) {
		const uint64_t O = mdl->nobs;
		const uint64_t F = mdl->nftr;
		info("* Compacting the model\n");
		mdl_compact(mdl);
		info("    %8"PRIu64" observations removed\n", O - mdl->nobs);
		info("    %8"PRIu64" features removed\n", F - mdl->nftr);
	}
	// And save the updated model
	info("* Save the model\n");
	FILE *file = stdout;
	if (mdl->opt->output != NULL) {
		file = fopen(mdl->opt->output, "w");
		if (file == NULL)
			pfatal("cannot open output model");
	}
	mdl_save(mdl, file);
	if (mdl->opt->output != NULL)
		fclose(file);
	info("* Done\n");
}

/*******************************************************************************
 * Entry point
 ******************************************************************************/
int main(int argc, char *argv[argc]) {
	// We first parse command line switchs
	opt_t opt = opt_defaults;
	opt_parse(argc, argv, &opt);
	// Next we prepare the model
	mdl_t *mdl = mdl_new(rdr_new(opt.maxent));
	mdl->opt = &opt;
	// And switch to requested mode
	switch (opt.mode) {
		case 0: dotrain(mdl); break;
		case 1: dolabel(mdl); break;
		case 2: dodump(mdl);  break;
		case 3: doupdt(mdl);  break;
	}
	// And cleanup
	mdl_free(mdl);
	return EXIT_SUCCESS;
}

